### PR TITLE
SDIT-2280 PERSON_IDENTIFIERS-UPDATED

### PR DESCRIPTION
**Pull request by andymarke**
Also refactor for telemetry success/error handling

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt [hunk 1]
```diff
@@ -47,6 +48,8 @@ import java.time.LocalDate
 import java.time.LocalDateTime
 import java.time.Period
 
+private fun telemetryOf(vararg pairs: Pair<String, Any>): MutableMap<String, Any> = mutableMapOf(*pairs)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt:51]: reduce the noise for the types e.g <String, Any> everywhere can be removed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt [hunk 2]
```diff
@@ -60,46 +63,46 @@ class ContactPersonSynchronisationService(
...
+    )
 
     if (event.doesOriginateInDps()) {
       telemetryClient.trackEvent(
         "contactperson-person-restriction-synchronisation-upserted-skipped",
         telemetry,
       )
     } else {
-      try {
+      val existingMapping = mappingApiService.getByNomisPersonRestrictionIdOrNull(nomisPersonRestrictionId = event.visitorRestrictionId)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt:78]: had to reorder for this UPSERT to get telemetry name right.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt [hunk 3]
```diff
@@ -127,29 +130,30 @@ class ContactPersonSynchronisationService(
...
-          val nomisContact = nomisPerson.contacts.find { it.id == event.contactId } ?: throw IllegalStateException("Contact ${event.contactId} for person ${event.personId} not found in NOMIS")
-          val dpsPrisonerContact = dpsApiService.createPrisonerContact(nomisContact.toDpsCreatePrisonerContactRequest(nomisPersonId = event.personId)).also {
-            telemetry["dpsPrisonerContactId"] = it.id
-          }
-          val mapping = PersonContactMappingDto(
-            nomisId = event.contactId,
-            dpsId = dpsPrisonerContact.id.toString(),
-            mappingType = PersonContactMappingDto.MappingType.NOMIS_CREATED,
-          )
+        track("contactperson-contact-synchronisation-created", telemetry) {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt:133]: same change everywhere - everything wrapped in a `track` call that outputs the correct telemetry depending on success/exception

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt [hunk 4]
```diff
@@ -630,35 +617,52 @@ class ContactPersonSynchronisationService(
...
-      telemetry,
-    )
+      telemetryOf("nomisPersonId" to event.personId, "dpsContactId" to event.personId, "nomisSequenceNumber" to event.identifierSequence)
+
+    if (event.doesOriginateInDps()) {
+      telemetryClient.trackEvent(
+        "contactperson-identifier-synchronisation-updated-skipped",
+        telemetry,
+      )
+    } else {
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt:649]: new functionality here

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt (hunk 1 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/contactperson/ContactPersonSynchronisationService.kt:51]: Prob worth adding this to the Telemetry class so everyone can use it?

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**